pandas/tests/reshape/test_from_dummies.py

```diff
@@ -0,0 +1,350 @@
+            r"Received 'base_category' of type: list"
+        ),
+    ):
+        from_dummies(dummies, base_category=["c", "d"])
+
+
+def test_from_dummies_no_prefix_multi_assignment():
+    dummies = DataFrame({"a": [1, 0, 1], "b": [0, 1, 1]})
+    with pytest.raises(
+        ValueError,
+        match=(
+            r"Dummy DataFrame contains multi-assignment\(s\); "
+            r"First instance in row: 2"
+        ),
+    ):
+        from_dummies(dummies)
+
+
+def test_from_dummies_no_prefix_contains_nan():
+    dummies = DataFrame({"a": [1, 0, 0], "b": [0, 1, np.nan]})
+    with pytest.raises(
+        ValueError, match=r"Dummy DataFrame contains NA value in column: 'b'"
+    ):
+        from_dummies(dummies)
+
+
+def test_from_dummies_contains_non_dummies():
+    dummies = DataFrame(
+        {"a": [1, 6, 3, 1], "b": [0, 1, 0, 2], "c": ["c1", "c2", "c3", "c4"]}
+    )
+    with pytest.raises(
+        TypeError,
+        match=r"Passed DataFrame contains non-dummy data",
+    ):
+        from_dummies(dummies)
+
+
+def test_from_dummies_with_prefix_basic(dummies_basic):
+    expected = DataFrame({"col1": ["a", "b", "a"], "col2": ["b", "a", "c"]})
+    result = from_dummies(dummies_basic, sep="_")
+    tm.assert_frame_equal(result, expected)
+
+
+def test_from_dummies_with_prefix_multiple_seperators():
+    dummies = DataFrame(
+        {
+            "col1_a": [1, 0, 1],
+            "col1_b": [0, 1, 0],
+            "col2-a": [0, 1, 0],
+            "col2-b": [1, 0, 1],
+        },
+    )
+    with pytest.raises(
+        ValueError,
+        match=(r"Separator not specified for column: col2-a"),
+    ):
+        from_dummies(dummies, sep="_")
+
+
+def test_from_dummies_with_prefix_sep_wrong_type(dummies_basic):
+
+    with pytest.raises(
+        TypeError,
+        match=(
+            r"Expected 'sep' to be of type 'str' or 'None'; "
+            r"Received 'sep' of type: list"
+        ),
+    ):
+        from_dummies(dummies_basic, sep=["_"])
+
+
+def test_from_dummies_with_prefix_contains_get_dummies_NaN_column():
+    dummies = DataFrame(
+        {
+            "col1_a": [1, 0, 0],
+            "col1_b": [0, 1, 0],
+            "col1_NaN": [0, 0, 1],
+            "col2_a": [0, 1, 0],
+            "col2_b": [0, 0, 0],
+            "col2_c": [0, 0, 1],
+            "col2_NaN": [1, 0, 0],
+        },
+    )
+    expected = DataFrame({"col1": ["a", "b", "NaN"], "col2": ["NaN", "a", "c"]})
+    result = from_dummies(dummies, sep="_")
+    tm.assert_frame_equal(result, expected)
+
+
+def test_from_dummies_with_prefix_contains_unassigned(dummies_with_unassigned):
+    with pytest.raises(
+        ValueError,
+        match=(
+            r"Dummy DataFrame contains unassigned value\(s\); "
+            r"First instance in row: 2"
+        ),
+    ):
+        from_dummies(dummies_with_unassigned, sep="_")
+
+
+def test_from_dummies_with_prefix_base_category_str(dummies_with_unassigned):
```

Comment:
what happens if I don't provide the base_category? 

*why* do we really need base_category (you prob mentioned once before )....

Reply:
> why do we really need base_category (you prob mentioned once before )....

It's common to drop a category when calling `get_dummies` (e.g. when using a model affected by collinearity), so this would ensure it's possible to round-trip

Reply:
> what happens if I don't provide the base_category?
> 

If the input data has unassigned values (which would be the candidates for the `base_category` value) and no `base_category` is provided, then it raises the error for containing unassigned entries (same as in `test_error_with_prefix_contains_unassigned`) as these values are not defined.